src/convert.cpp

```diff
@@ -30,14 +31,97 @@ const auto tof_to_s =
 const auto J_to_meV =
     units::meV /
     boost::units::quantity<boost::units::si::energy>(1.0 * units::meV);
+const auto m_to_angstrom =
+    boost::units::quantity<boost::units::si::length>(1.0 * units::angstrom) /
+    units::angstrom;
 // In tof-to-energy conversions we *divide* by time-of-flight (squared), so the
 // tof_to_s factor is in the denominator.
 const auto tofToEnergyPhysicalConstants =
     0.5 * boost::units::si::constants::codata::m_n * J_to_meV /
     (tof_to_s * tof_to_s);
+const auto tofToDSpacingPhysicalConstants =
+    2.0 * boost::units::si::constants::codata::m_n * m_to_angstrom /
+    (boost::units::si::constants::codata::h * tof_to_s);
 
 namespace neutron {
 namespace tof {
+Dataset tofToDSpacing(const Dataset &d) {
+  if (d.contains(Coord::Ei) || d.contains(Coord::Ef))
+    throw std::runtime_error(
+        "Dataset contains Coord::Ei or Coord::Ef. "
+        "However, conversion to Dim::DSpacing is currently "
+        "only supported for elastic scattering.");
+
+  // 1. Compute conversion factor
+  const auto &compPos = d.get(Coord::ComponentInfo)[0](Coord::Position);
+  // TODO Need a better mechanism to identify source and sample.
+  const auto &sourcePos = compPos(Dim::Component, 0);
+  const auto &samplePos = compPos(Dim::Component, 1);
+  const auto l1 = norm(sourcePos - samplePos);
+  const auto specPos = getSpecPos(d);
+
+  // l_total = l1 + l2
+  auto conversionFactor(norm(specPos - samplePos) + l1);
+
+  conversionFactor *= tofToDSpacingPhysicalConstants;
+
+  // sin(scattering_angle)
+  auto beam = samplePos - sourcePos;
+  beam /= norm(beam);
+  auto scattered = specPos - samplePos;
+  scattered /= norm(scattered);
+
+  // TODO Need `dot` for `Variable`. The following block should just be
+  // conversionFactor *= sqrt(0.5 * (1.0 - dot(beam, scattered)))
+
+  std::vector<double> sinThetaData(scattered.size());
+  const auto &beamVec = beam.span<Eigen::Vector3d>()[0];
+  const auto scatteredVec = scattered.span<Eigen::Vector3d>();
+  // Using
+  //   cos(2 theta) = 1 - 2 sin^2(theta)
+  // and
+  //   v1 dot v2 = norm(v1) norm(v2) cos(alpha).
+  std::transform(scatteredVec.begin(), scatteredVec.end(), sinThetaData.begin(),
+                 [&](const Eigen::Vector3d &scattered) {
+                   return sqrt(0.5 * (1.0 - beamVec.dot(scattered)));
+                 });
+  const auto sinTheta =
+      makeVariable<double>(Data::Value, scattered.dimensions(), sinThetaData);
+  conversionFactor *= sinTheta;
+
+  // 2. Transform coordinate
+  Dataset converted;
+  const auto &coord = d(Coord::Tof);
+  auto coordDims = coord.dimensions();
+  coordDims.relabel(coordDims.index(Dim::Tof), Dim::DSpacing);
+  // The reshape is to remap the dimension label.
+  converted.insert(Coord::DSpacing,
+                   coord.reshape(coordDims) / conversionFactor);
+
+  // 3. Transform variables
```

Comment:
Not something that should be done now, but i'm wondering if some of the conversions code could be put in common in the future. The different conversions at the moment seem to all have a similar structure:
1. Define some conversion factors
2. Convert the coordinate
3. Transform the variables

Especially number 3 looks similar, but I can see that one may need to do things slightly differently for different conversions, so I'm not sure how easy it would be to make some sort of generic conversion operation.
I don't know, maybe this wouldn't be useful, because each conversion would be too different anyway.

Reply:
Yes, this will certainly be done.